core/src/main/java/org/apache/spark/SparkFirehoseListener.java

```diff
@@ -140,6 +140,16 @@ public void onBlockUpdated(SparkListenerBlockUpdated blockUpdated) {
   }
 
   @Override
+  public void onSpeculativeTaskSubmitted(SparkListenerSpeculativeTaskSubmitted speculativeTask) {
+    onEvent(speculativeTask);
+  }
+
+  @Override
+  public void onExtraExecutorNeeded() {
```

Comment:
When there is one executor left but with multiple cpu cores, the task is running on that executor. But due to locality requirements, speculative jobs cannot launch on the same host. We will have to request for one extra executor. Thats is what this event is for.